From 98c3512b43df80b78af0142c5b8d1ea0b9154f96 Mon Sep 17 00:00:00 2001
From: Zumeng Chen <zumeng.chen@windriver.com>
Date: Mon, 13 Mar 2017 16:08:42 +0800
Subject: [PATCH 1/2] yaffs: fix potential memory leakage and codes redundant

put_mtd_device should fix the following calltrace

unreferenced object 0xe9db0180 (size 128):
  comm "mount", pid 2742, jiffies 284437 (age 176.640s)
  hex dump (first 32 bytes):
    80 01 db e9 80 01 db e9 00 c1 d6 eb 00 77 c9 eb .............w..
    00 00 00 00 00 00 00 00 01 00 00 00 00 00 00 00 ................
  backtrace:
    [<c012909c>] kmem_cache_alloc_trace+0x194/0x278
    [<c034ce50>] yaffs_internal_read_super.isra.16+0x2b4/0x724
    [<c034d2e8>] yaffs2_internal_read_super_mtd+0x28/0x38
    [<c013ccbc>] mount_bdev+0x12c/0x190
    [<c034a020>] yaffs2_mount+0x24/0x30
    [<c013d71c>] mount_fs+0x80/0x178
    [<c0157948>] vfs_kern_mount+0x58/0xec
    [<c015a53c>] do_mount+0x8f4/0xa08
    [<c015a8a0>] SyS_mount+0x94/0xc8
    [<c000e320>] ret_fast_syscall+0x0/0x30
    [<ffffffff>] 0xffffffff

Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 fs/yaffs2/yaffs_vfs.c |    4 +---
 1 files changed, 1 insertions(+), 3 deletions(-)

diff --git a/fs/yaffs2/yaffs_vfs.c b/fs/yaffs2/yaffs_vfs.c
index 04fd1d8..a276e69 100644
--- a/fs/yaffs2/yaffs_vfs.c
+++ b/fs/yaffs2/yaffs_vfs.c
@@ -2896,15 +2896,13 @@ static struct super_block *yaffs_internal_read_super(int yaffs_version,
 	context = kmalloc(sizeof(struct yaffs_linux_context), GFP_KERNEL);
 
 	if (!dev || !context) {
+		put_mtd_device(mtd);
 		if (dev)
 			kfree(dev);
 		if (context)
 			kfree(context);
 		dev = NULL;
 		context = NULL;
-	}
-
-	if (!dev) {
 		/* Deep shit could not allocate device structure */
 		yaffs_trace(YAFFS_TRACE_ALWAYS,
 			"yaffs_read_super: Failed trying to allocate struct yaffs_dev."
-- 
1.7.5.4

